### PR TITLE
feat: add Smithy model-driven conformance testing

### PR DESCRIPTION
**Pull request by vieiralucas**
## Summary

- Add conformance testing tool that automatically measures AWS API coverage by probing fakecloud with requests generated from official Smithy models
- Parses all 983 operations across 13 services, generates ~15k test variants using 5 strategies (boundary values, enum exhaustion, optionals permutation, model examples, negative testing)
- Validates response shapes against Smithy model output definitions
- Current results: **76% pass rate** across 15k+ variants with shape validation
- Includes `#[test_action]` proc macro with model checksums for future handwritten conformance tests (Level 2)

## Test plan

- [x] `cargo test -p fakecloud-conformance --lib` — 17 unit tests pass
- [x] `cargo clippy -p fakecloud-conformance -p fakecloud-conformance-macros -- -D warnings` — clean
- [x] `cargo run -p fakecloud-conformance -- run` — full suite completes in seconds
- [x] `cargo run -p fakecloud-conformance -- operations` — lists all 983 operations
- [x] `cargo run -p fakecloud-conformance -- checksums` — prints checksums for all operations

---
## Summary by cubic
Adds Smithy model–driven conformance testing to measure AWS API coverage and validate responses against official models. Includes a CLI, a CI workflow with a markdown table summary, a macro for model‑anchored tests, and follow‑up fixes that harden validation, generation, and CI output.

- **New Features**
  - `fakecloud-conformance` CLI loads 13 AWS Smithy models (983 operations) and probes a running `fakecloud` using five strategies (boundary, enum, optionals, examples, negative).
  - Validates response shapes, classifies outcomes, and outputs text/JSON reports. Initial run shows a 76% pass rate.
  - `fakecloud-conformance-macros` adds `#[test_action(...)]` with compile‑time model lookups and operation checksums.
  - Adds `.github/workflows/conformance.yml` to build, run probes on PRs/main/weekly, post a markdown table summary, and upload artifacts.
  - Adds `scripts/update-aws-models.sh` and checks in Smithy models under `aws-models/`.

- **Bug Fixes**
  - Robustness: ensure server cleanup with ChildGuard; precise int checks (`is_i64`/`is_u64`); require `expected_code` for error expectations; treat empty probe results as not_implemented; fix multibyte truncate panic.
  - Generation/validation: test empty strings for min=1; guard integer overflow in negatives; enforce exactly‑one‑member for unions; dedupe examples; append (not overwrite) enum constraints; guard CI summary output; skip duplicate `all_fields` when fewer than two optionals; allow empty string boundary when min=0; merge member‑level and target‑level constraints for boundary generation; plus clippy unused var prefix and dead test removal.

<sup>Written for commit 4e0e0edc5d0531b4d11007908e54bc89d13e3c1d. Summary will update on new commits.</sup>

### CODE REVIEW

File: crates/fakecloud-conformance/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    format: &str,
+    endpoint: Option<String>,
+) {
+    let models = load_models(models_dir);
+
+    let filter: Option<Vec<String>> =
+        services_filter.map(|s| s.split(',').map(|s| s.trim().to_string()).collect());
+
+    // Start fakecloud or connect to existing
+    let (endpoint, _server) = if let Some(ep) = endpoint {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/main.rs:145]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/shape_validator.rs [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+        ShapeType::String { .. } | ShapeType::Enum { .. } => {
+            if !value.is_string() {
+                violations.push(ShapeViolation::WrongType {
+                    path: path.to_string(),
+                    expected: "string".to_string(),
+                    got: json_type_name(value).to_string(),
+                });
+            }
+        }
+        ShapeType::Integer | ShapeType::Long | ShapeType::IntEnum { .. } => {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/shape_validator.rs:244]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/probe.rs [hunk 3]
```diff
@@ -0,0 +1,838 @@
...
+        }
+        Expectation::AnyError => {
+            if http_status >= 400 {
+                ProbeStatus::Pass
+            } else {
+                ProbeStatus::UnexpectedResult(format!("Expected error, got HTTP {}", http_status))
+            }
+        }
+        Expectation::Error(expected_code) => {
+            if body.contains(expected_code) || http_status >= 400 {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/probe.rs]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/probe.rs [hunk 4]
```diff
@@ -0,0 +1,838 @@
...
+            status: ProbeStatus::Crash(format!("HTTP 500: {}", truncate(body, 200))),
+            http_status,
+            response_body: body.to_string(),
+            duration_ms,
+        };
+    }
+
+    let status = match expectation {
+        Expectation::Success => {
+            if (200..500).contains(&http_status) {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/probe.rs:755]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/generators/negative.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        let shape = match model.shapes.get(&member.target) {
+            Some(s) => s,
+            None => continue,
+        };
+
+        let traits = &shape.traits;
+
+        // String too short (below min length)
+        if let Some(min) = traits.length_min {
+            if min > 1 {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/negative.rs]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/generators/negative.rs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                }
+                _ => {}
+            }
+        }
+
+        // Integer above max range
+        if let Some(max) = traits.range_max {
+            match &shape.shape_type {
+                ShapeType::Integer | ShapeType::Long => {
+                    let above = (max as i64) + 1;
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/negative.rs]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/generators/negative.rs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                    });
+                }
+            }
+        }
+
+        // Integer below min range
+        if let Some(min) = traits.range_min {
+            match &shape.shape_type {
+                ShapeType::Integer | ShapeType::Long => {
+                    let below = (min as i64) - 1;
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/negative.rs]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/report.rs [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        summary.total_operations += total_ops;
+
+        let mut operations = Vec::new();
+        let mut op_names: Vec<_> = ops.keys().cloned().collect();
+        op_names.sort();
+
+        for op_name in op_names {
+            let probe_results = &ops[&op_name];
+
+            let not_implemented = probe_results
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/report.rs]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/shape_validator.rs [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+        ShapeType::List { member_target } => {
+            validate_list(model, member_target, value, path, depth, violations);
+        }
+        ShapeType::Map { value_target, .. } => {
+            validate_map(model, value_target, value, path, depth, violations);
+        }
+        ShapeType::Union { members } => {
+            // A union should be an object with exactly one key matching a member name.
+            // For validation purposes, we treat it like a structure but don't require fields.
+            if let Value::Object(obj) = value {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/shape_validator.rs:201]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/generators/mod.rs [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+
+    // Strategy 5: Examples from model
+    let op_shape_id = format!("{}#{}", model.service_name, operation_name);
+    // Try to find examples on the operation shape
+    if let Some(op_shape) = model.shapes.get(&op_shape_id) {
+        variants.extend(examples::generate(&op_shape.traits));
+    }
+    // Also check the operation's own examples
+    for (shape_id, shape) in &model.shapes {
+        if shape_id.ends_with(&format!("#{}", operation_name))
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/mod.rs]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/probe.rs [hunk 11]
```diff
@@ -0,0 +1,838 @@
...
+        .replace('=', "%3D")
+        .replace('+', "%2B")
+        .replace(' ', "%20")
+}
+
+fn truncate(s: &str, max: usize) -> &str {
+    if s.len() <= max {
+        s
+    } else {
+        &s[..max]
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/probe.rs]: Fixed in eaae798.

File: .github/workflows/conformance.yml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        run: |
+          cargo run -p fakecloud-conformance -- run --format text 2>&1 | tee conformance-report.txt
+          cargo run -p fakecloud-conformance -- run --format json > conformance-report.json
+
+      - name: Post summary
+        if: always()
+        run: |
+          echo '## Conformance Report' >> $GITHUB_STEP_SUMMARY
+          echo '```' >> $GITHUB_STEP_SUMMARY
+          tail -20 conformance-report.txt >> $GITHUB_STEP_SUMMARY
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply .github/workflows/conformance.yml]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/checksum.rs [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+            key_target,
+            value_target,
+        } => {
+            collect_shape_tree(model, key_target, collected, visited);
+            collect_shape_tree(model, value_target, collected, visited);
+        }
+        ShapeType::String {
+            enum_values: Some(values),
+        } => {
+            canonical.constraints = format!(
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/checksum.rs]: Fixed in eaae798.

File: .github/workflows/conformance.yml [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+          cargo run -p fakecloud-conformance -- run --format json > conformance-report.json
+
+      - name: Post summary
+        if: always()
+        run: |
+          echo '## Conformance Report' >> $GITHUB_STEP_SUMMARY
+          echo '' >> $GITHUB_STEP_SUMMARY
+          echo '| Service | Handled | Passing | Variants |' >> $GITHUB_STEP_SUMMARY
+          echo '|---------|---------|---------|----------|' >> $GITHUB_STEP_SUMMARY
+          grep -E "^[a-z].*operations handled" conformance-report.txt | while read -r line; do
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply .github/workflows/conformance.yml:42]: Fixed in eaae798.

File: crates/fakecloud-conformance/src/generators/optionals.rs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    // Required-only variant
+    let required_input = build_required_input(model, input_shape_id, overrides);
+    variants.push(TestVariant {
+        name: "required_only".to_string(),
+        strategy: Strategy::Optionals,
+        input: required_input,
+        expectation: Expectation::Success,
+    });
+
+    // All-fields variant
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/optionals.rs]: Acknowledged — the duplication is minor (2 variants vs 1 for shapes with no optionals) and doesn't affect correctness. Will revisit if the variant count becomes a problem.

File: crates/fakecloud-conformance/src/generators/boundary.rs [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+
+        if !has_length && !has_range {
+            continue;
+        }
+
+        // String length boundaries
+        if has_length {
+            if let ShapeType::String { .. } = &shape.shape_type {
+                if let Some(min) = traits.length_min {
+                    let min = min.max(1) as usize;
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/boundary.rs]: Fair point. In practice no AWS shapes use min=0 for strings (empty strings are never valid API inputs), so this is a theoretical edge case. Will fix if we encounter it.

File: crates/fakecloud-conformance/src/generators/boundary.rs [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+            .get(&member.target)
+            .or_else(|| smithy::prelude_shape_type(&member.target).map(|_| &*PLACEHOLDER_SHAPE));
+
+        let shape = match target_shape {
+            Some(s) => s,
+            None => continue,
+        };
+
+        // Get constraints from the shape's traits AND the member's own traits
+        let traits = &shape.traits;
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/generators/boundary.rs]: Good catch. Member-level constraints are relatively rare in AWS models (most constraints are on the target shape). Adding this would require merging member traits with target shape traits throughout the generator — deferring to a follow-up.

File: crates/fakecloud-conformance-macros/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,592 @@
...
+             The Smithy model has changed — update the checksum or the conformance test.",
+            args.service, args.action, args.checksum, actual_checksum
+        );
+        return syn::Error::new(proc_macro2::Span::call_site(), msg)
+            .to_compile_error()
+            .into();
+    }
+
+    // Pass through the original function unchanged
+    let output = quote! {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance-macros/src/lib.rs:588]: Proc macros can't emit cargo:rerun-if-changed directives. The models are checked into git and change infrequently (manual script run). In practice, any model update comes with a code change that triggers recompilation. If this becomes a real issue we can add a build.rs to the consuming crate.

File: crates/fakecloud-conformance/src/smithy.rs [hunk 19]
```diff
@@ -0,0 +1,688 @@
...
+        "smithy.api#PrimitiveFloat" => Some(ShapeType::Float),
+        "smithy.api#PrimitiveDouble" => Some(ShapeType::Double),
+        "smithy.api#PrimitiveShort" => Some(ShapeType::Integer),
+        "smithy.api#PrimitiveByte" => Some(ShapeType::Integer),
+        _ => None,
+    }
+}
+
+/// Resolve a shape ID to its Shape, handling prelude types.
+pub fn resolve_shape<'a>(model: &'a ServiceModel, shape_id: &str) -> Option<&'a Shape> {
```
[the remark this answers is not in the document]
Comment (vieiralucas) [reply crates/fakecloud-conformance/src/smithy.rs:538]: Acknowledged — the function is unused and behind `#![allow(dead_code)]`. It exists for future use by the Level 2 audit system. Will clean up the docstring or remove it if it stays unused.
